commit 07a0ffb6c26be63eeb80bcf1d17a38b9e7893a3f
Author: Dan Bradley <dan@hep.wisc.edu>
Date:   Fri Jun 15 11:41:04 2012 -0500

    Check that the daemon's GSI credential matches its hostname. #1605
    
    To get the old behavior in which the hostname is not checked,
    set GSI_SKIP_HOST_CHECK=true.

diff --git a/src/condor_includes/condor_auth_x509.h b/src/condor_includes/condor_auth_x509.h
index 6143a01..bdbc545 100644
--- a/src/condor_includes/condor_auth_x509.h
+++ b/src/condor_includes/condor_auth_x509.h
@@ -118,6 +118,12 @@ class Condor_Auth_X509 : public Condor_Auth_Base {
 	@return TRUE if valid FALSE if not */
     bool gss_is_valid();
 
+	/** See if host name in peer's certificate matches what we
+		are connecting to.
+		@return true if valid; false if not
+	 */
+	bool CheckServerName(char const *fqh,char const *ip,ReliSock *sock,CondorError *errstack);
+
     /** A specialized function that is needed for secure personal
         condor. When schedd and the user are running under the
         same userid we would still want the authentication process
@@ -136,6 +142,7 @@ class Condor_Auth_X509 : public Condor_Auth_Base {
     //------------------------------------------
     gss_cred_id_t       credential_handle;
     gss_ctx_id_t        context_handle ;
+    gss_name_t          m_gss_server_name;
     int                 token_status;
     //X509_Credential *   my_credential;
     OM_uint32	        ret_flags ;
diff --git a/src/condor_includes/condor_error_codes.h b/src/condor_includes/condor_error_codes.h
index fba29dc..2b3ee6c 100644
--- a/src/condor_includes/condor_error_codes.h
+++ b/src/condor_includes/condor_error_codes.h
@@ -52,6 +52,7 @@ const int GSI_ERR_AUTHENTICATION_FAILED = 5004;
 const int GSI_ERR_COMMUNICATIONS_ERROR = 5005;
 const int GSI_ERR_UNAUTHORIZED_SERVER = 5006;
 const int GSI_ERR_NO_VALID_PROXY = 5007;
+const int GSI_ERR_DNS_CHECK_ERROR = 5008;
 
 const int CEDAR_ERR_CONNECT_FAILED = 6001;
 const int CEDAR_ERR_EOM_FAILED = 6002;
diff --git a/src/condor_io/condor_auth_x509.cpp b/src/condor_io/condor_auth_x509.cpp
index 3940144..4475623 100644
--- a/src/condor_io/condor_auth_x509.cpp
+++ b/src/condor_io/condor_auth_x509.cpp
@@ -41,7 +41,7 @@ extern "C" {
 #define USER_NAME_MAX 256
 
 const char STR_DAEMON_NAME_FORMAT[]="$$(FULL_HOST_NAME)";
-StringList * getDaemonList(ReliSock * sock);
+StringList * getDaemonList(char const *param_name,char const *fqh);
 
 
 #ifdef WIN32
@@ -58,6 +58,7 @@ Condor_Auth_X509 :: Condor_Auth_X509(ReliSock * sock)
     : Condor_Auth_Base (sock, CAUTH_GSI),
       credential_handle(GSS_C_NO_CREDENTIAL),
       context_handle   (GSS_C_NO_CONTEXT),
+      m_gss_server_name(NULL),
       token_status     (0),
       ret_flags        (0)
 {
@@ -84,6 +85,11 @@ Condor_Auth_X509 ::  ~Condor_Auth_X509()
         OM_uint32 major_status = 0; 
         gss_release_cred(&major_status, &credential_handle);
     }
+
+	if( m_gss_server_name != NULL ) {
+		OM_uint32 major_status = 0;
+		gss_release_name( &major_status, &m_gss_server_name );
+	}
 }
 
 int Condor_Auth_X509 :: authenticate(const char * /* remoteHost */, CondorError* errstack)
@@ -487,18 +493,16 @@ int Condor_Auth_X509::nameGssToLocal(const char * GSSClientname)
 	return 1;
 }
 
-StringList * getDaemonList(ReliSock * sock)
+StringList * getDaemonList(char const *param_name,char const *fqh)
 {
     // Now, we substitu all $$FULL_HOST_NAME with actual host name, then
     // build a string list, then do a search to see if the target is 
     // in the list
-    char * daemonNames = param( "GSI_DAEMON_NAME" );
-	MyString fqh_str = get_hostname(sock->peer_addr());
-    const char * fqh  = fqh_str.Value();
+    char * daemonNames = param( param_name );
     char * entry       = NULL;
 
 	if (!daemonNames) {
-		daemonNames = strdup("*");
+		return NULL;
 	}
 
     StringList * original_names = new StringList(daemonNames, ",");
@@ -549,7 +553,6 @@ char * Condor_Auth_X509::get_server_info()
     OM_uint32	major_status = 0;
     OM_uint32	minor_status = 0;            
     OM_uint32   lifetime, flags;
-    gss_name_t  target = NULL;
     gss_OID     mech, name_type;
     gss_buffer_desc name_buf;
     char *      server = NULL;
@@ -558,7 +561,7 @@ char * Condor_Auth_X509::get_server_info()
     major_status = gss_inquire_context(&minor_status,
                                        context_handle,
                                        NULL,    
-                                       &target,
+                                       &m_gss_server_name,
                                        &lifetime,
                                        &mech, 
                                        &flags, 
@@ -568,23 +571,21 @@ char * Condor_Auth_X509::get_server_info()
         dprintf(D_SECURITY, "Unable to obtain target principal name\n");
         return NULL;
     }
-    else {
-        major_status = gss_display_name(&minor_status,
-                                             target,
-                                             &name_buf,
-                                             &name_type);
-		gss_release_name( &minor_status, &target );
-		if( major_status != GSS_S_COMPLETE) {
-            dprintf(D_SECURITY, "Unable to convert target principal name\n");
-            return NULL;
-        }
-        else {
-            server = new char[name_buf.length+1];
-            memset(server, 0, name_buf.length+1);
-            memcpy(server, name_buf.value, name_buf.length);
-			gss_release_buffer( &minor_status, &name_buf );
-        }
-    }
+
+	major_status = gss_display_name(&minor_status,
+									m_gss_server_name,
+									&name_buf,
+									&name_type);
+	if( major_status != GSS_S_COMPLETE) {
+		dprintf(D_SECURITY, "Unable to convert target principal name\n");
+		return NULL;
+	}
+
+	server = new char[name_buf.length+1];
+	memset(server, 0, name_buf.length+1);
+	memcpy(server, name_buf.value, name_buf.length);
+	gss_release_buffer( &minor_status, &name_buf );
+
     return server;
 }   
 
@@ -774,17 +775,34 @@ int Condor_Auth_X509::authenticate_client_gss(CondorError* errstack)
 			}
 		}
 
-        StringList * daemonNames = getDaemonList(mySock_);
+        std::string fqh = convert_ipaddr_to_hostname(mySock_->peer_addr());
+        StringList * daemonNames = getDaemonList("GSI_DAEMON_NAME",fqh.c_str());
 
         // Now, let's see if the name is in the list, I am not using
         // anycase here, so if the host name and what we are looking for
         // are in different cases, then we will run into problems.
-        status = daemonNames->contains_withwildcard(server) == TRUE? 1 : 0;
+		if( daemonNames ) {
+			status = daemonNames->contains_withwildcard(server) == TRUE? 1 : 0;
+		}
+		else {
+			status = 1;
+		}
+
+		bool host_check_failed = false;
+		if( !status || !daemonNames ) {
+				// DN not found in GSI_DAEMON_NAME
+
+			host_check_failed = !CheckServerName(fqh.c_str(),mySock_->peer_ip_str(),mySock_,errstack);
+
+			if( host_check_failed ) {
+				status = 0;
+			}
+		}
 
         if (status) {
             dprintf(D_SECURITY, "valid GSS connection established to %s\n", server);            
         }
-        else {
+        else if( !host_check_failed ) {
 			errstack->pushf("GSI", GSI_ERR_UNAUTHORIZED_SERVER,
 					"Failed to authenticate because the subject '%s' is not currently trusted by you.  "
 					"If it should be, add it to GSI_DAEMON_NAME in the condor_config, "
@@ -807,6 +825,89 @@ int Condor_Auth_X509::authenticate_client_gss(CondorError* errstack)
     return (status == 0) ? FALSE : TRUE;
 }
 
+bool Condor_Auth_X509::CheckServerName(char const *fqh,char const *ip,ReliSock *sock,CondorError *errstack)
+{
+	if( param_boolean("GSI_SKIP_HOST_CHECK",false) ) {
+		return true;
+	}
+
+	char const *server_dn = getAuthenticatedName();
+	if( !server_dn ) {
+		std::string msg;
+		sprintf(msg,"Failed to find certificate DN for server on GSI connection to %s",ip);
+		errstack->push("GSI", GSI_ERR_DNS_CHECK_ERROR, msg.c_str());
+		return false;
+	}
+
+	std::string skip_check_pattern;
+	if( param(skip_check_pattern,"GSI_SKIP_HOST_CHECK_CERT_REGEX") ) {
+		Regex re;
+		const char *errptr=NULL;
+		int erroffset=0;
+		if( !re.compile(skip_check_pattern.c_str(),&errptr,&erroffset) ) {
+			dprintf(D_ALWAYS,"GSI_SKIP_HOST_CHECK_CERT_REGEX is not a valid regular expression: %s\n",skip_check_pattern.c_str());
+			return false;
+		}
+		if( re.match(server_dn) ) {
+			return true;
+		}
+	}
+
+	ASSERT( errstack );
+	ASSERT( m_gss_server_name );
+	ASSERT( ip );
+	if( !fqh || !fqh[0] ) {
+		std::string msg;
+		sprintf(msg,"Failed to look up host address for GSI connection on %s to server with DN %s.  Perhaps this is not a host certificate?  This server name check can be bypassed by either adding the DN to GSI_DAEMON_NAME or GSI_SKIP_HOST_CHECK_CERT_REGEX to make an exception, or by disabling all host name checks by setting GSI_SKIP_HOST_CHECK=true.",ip,server_dn);
+		errstack->push("GSI", GSI_ERR_DNS_CHECK_ERROR, msg.c_str());
+		return false;
+	}
+
+	std::string connect_name;
+	gss_buffer_desc gss_connect_name_buf;
+	gss_name_t gss_connect_name;
+	OM_uint32 major_status = 0;
+	OM_uint32 minor_status = 0;
+
+	sprintf(connect_name,"%s/%s",fqh,sock->peer_ip_str());
+
+	gss_connect_name_buf.value = strdup(connect_name.c_str());
+	gss_connect_name_buf.length = connect_name.size()+1;
+
+	major_status = gss_import_name(&minor_status,
+								   &gss_connect_name_buf,
+								   GLOBUS_GSS_C_NT_HOST_IP,
+								   &gss_connect_name);
+
+	free( gss_connect_name_buf.value );
+
+	if( major_status != GSS_S_COMPLETE ) {
+		std::string comment;
+		sprintf(comment,"Failed to create gss connection name data structure for %s.\n",connect_name.c_str());
+		print_log( major_status, minor_status, 0, comment.c_str() );
+		return false;
+	}
+
+	int name_equal = 0;
+	major_status = gss_compare_name( &minor_status,
+									 m_gss_server_name,
+									 gss_connect_name,
+									 &name_equal );
+
+	gss_release_name( &major_status, &gss_connect_name );
+
+	if( !name_equal ) {
+		std::string msg;
+		sprintf(msg,"We are trying to connect to a daemon with certificate DN (%s), but the host name in the certificate does not match any DNS name associated with the host to which we are connecting (host name is '%s', IP is '%s', Condor connection address is '%s').  Check that DNS is correctly configured.  If you wish to use a daemon certificate that does not match the daemon's host name, either add the DN to GSI_DAEMON_NAME or GSI_SKIP_HOST_CHECK_CERT_REGEX to make an exception, or disable all host name checks by setting GSI_SKIP_HOST_CHECK=true.\n",
+				server_dn,
+				fqh,
+				ip,
+				sock->peer_description() );
+		errstack->push("GSI", GSI_ERR_DNS_CHECK_ERROR, msg.c_str());
+	}
+	return name_equal != 0;
+}
+
 int Condor_Auth_X509::authenticate_server_gss(CondorError* errstack)
 {
     char *    GSSClientname;
